Scripts/modelsystem.py

```diff
@@ -241,23 +244,28 @@ def run_iteration(self, previous_iter_impedance, iteration=None):
         self._add_internal_demand(previous_iter_impedance, iteration=="last")
 
         # Calculate external demand
-        trip_sum = {}
         for mode in param.external_modes:
             if mode == "truck":
                 int_demand = self.trucks.matrix.sum(0) + self.trucks.matrix.sum(1)
             elif mode == "trailer_truck":
                 int_demand = self.trailer_trucks.matrix.sum(0) + self.trailer_trucks.matrix.sum(1)
             else:
                 int_demand = self._sum_trips_per_zone(mode)
-                trip_sum[mode] = int_demand.sum()
             ext_demand = self.em.calc_external(mode, int_demand)
             self.dtm.add_demand(ext_demand)
+
+        # Calculate trips and mode shares
+        trip_sum = {}
+        for mode in self.travel_modes:
+            trip_sum[mode] = self._sum_trips_per_zone(mode)
         sum_all = sum(trip_sum.values())
-        # ATM, these mode shares are for car and transit
-        # for the whole model area
         mode_share = {}
         for mode in trip_sum:
-            mode_share[mode] = trip_sum[mode] / sum_all
+            self.resultdata.print_data(trip_sum[mode], "origins_demand.txt", 
+                self.zdata_base.zone_numbers, mode)
+            self.resultdata.print_data(trip_sum[mode] / sum_all, "origins_shares.txt", 
```

Comment:
Wait a minute, this is dividing the mode specific origin demand by demand for **all** zones.

Reply:
`sum_all ` should be vector (shape same as `trip_sum[mode])` containing sum for all modes. Sum over all zones is calculated on row 268.

Reply:
Yes, you are right, my bad!